### PR TITLE
Reduce state from arbitrary streams

### PR DESCRIPTION
**Pull request by totokaka**
This is what we have called "data reducers" in ardoq-front. I think everything should be clear from the commit messages and the code.

### CODE REVIEW

File: src/reducer.spec.ts [hunk 1]
```diff
@@ -39,35 +44,86 @@ test('reducer should store reducer function', t => {
...
-    m.expect(source.pipe(combineReducers(1, [alwaysReset]))).toBeObservable(
-      expected
+    m.expect(action$.pipe(combineReducers(1, [alwaysReset]))).toBeObservable(
+      expected$
     );
   })
 );
+
+test(
+  'combineReducers should not emit before after the seed stream does',
```
Comment (anton164) [inline src/reducer.spec.ts]:
```suggestion
  'combineReducers should not emit and ignore actions until the seed stream emits a value',
```

^ would this make it a bit more clear?

File: src/reducer.ts [hunk 2]
```diff
@@ -108,50 +139,120 @@ type ReducerCreator = {
...
+  type InitializedAccumulator = {
+    initialized: true;
+    reduced: boolean;
+    state: State;
+  };
+  type Accumulator = { initialized: false } | InitializedAccumulator;
+
+  const seed$ = wrapInObservable(seed).pipe(
+    take(1),
+    map(payload => [SEED_ORIGIN, payload] as ScanInput)
```
Comment (anton164) [inline src/reducer.ts]:
```suggestion
    map(initialState => [SEED_ORIGIN, initialState] as ScanInput)
```

File: src/reducer.ts [hunk 3]
```diff
@@ -108,50 +139,120 @@ type ReducerCreator = {
...
+        return {
+          initialized: true,
+          reduced: true,
+          state: reducerFn(accumulator.state, packet[1]),
+        };
+      },
+      { initialized: false } as Accumulator
+    ),
+    filter(
+      (acc): acc is InitializedAccumulator => acc.initialized && acc.reduced
```
Comment (anton164) [inline src/reducer.ts]: Does this mean that the combine reducers stream doesn't emit until it's reduced once?

Shouldn't it also emit once it's been initialized?

File: src/reducer.ts [hunk 4]
```diff
@@ -108,50 +139,120 @@ type ReducerCreator = {
...
-      const RegisteredReducer = reducersByActionType.get(type);
-      if (RegisteredReducer === undefined) {
-        // This shouldn't be possible
-        return state;
-      }
+    ofType(...actionReducers.flatMap(reducerFn => reducerFn.trigger.actions)),
+    map(action => [ACTION_ORIGIN, action] as ScanInput),
+    merge(seed$, ...source$s),
+    scan(
+      (accumulator, packet) => {
```
Comment (anton164) [inline src/reducer.ts]: Very elegant - reads easily! 👍

File: src/reducer.ts [hunk 5]
```diff
@@ -108,50 +139,120 @@ type ReducerCreator = {
...
     reducer: Reducer<State, VoidPayload>
   ): RegisteredReducer<State, VoidPayload>;
+
+  <State, Payload>(
+    source$: Observable<Payload>,
+    reducer: Reducer<State, Payload>
+  ): RegisteredReducer<State, Payload>;
 };
 
 export const reducer: ReducerCreator = <State>(
```
Comment (anton164) [inline src/reducer.ts]: Great idea to overload this and just treat this as two types of reducers with either an action or a stream payload :)

File: src/reducer.ts [hunk 6]
```diff
@@ -1,27 +1,58 @@
...
 
+const isActionReducer = <State, Payload>(
+  reducerFn: RegisteredReducer<State, Payload>
+): reducerFn is RegisteredActionReducer<State, Payload> =>
+  'actions' in reducerFn.trigger;
+
+const isStreamReducer = <State, Payload>(
+  reducerFn: RegisteredReducer<State, Payload>
+): reducerFn is RegisteredStreamReducer<State, Payload> =>
+  'source$' in reducerFn.trigger;
```
Comment (chriskr) [inline src/reducer.ts]: Use const for 'actions' and 'sources$' ?

File: src/reducer.ts [hunk 7]
```diff
@@ -108,50 +139,120 @@ type ReducerCreator = {
...
-        // This shouldn't be possible
-        return state;
-      }
+    ofType(...actionReducers.flatMap(reducerFn => reducerFn.trigger.actions)),
+    map(action => [ACTION_ORIGIN, action] as ScanInput),
+    merge(seed$, ...source$s),
+    scan(
+      (accumulator, packet) => {
+        if (packet[0] === SEED_ORIGIN) {
+          return { initialized: true, reduced: false, state: packet[1] };
```
Comment (chriskr) [inline src/reducer.ts]: If if this is internal I think it would make sense to destruct packet to named variables

File: src/reducer.ts (hunk 3 above)
Comment (chriskr) [reply src/reducer.ts]: Yes, I would think the same?

Ok, I see, that correspond to how scan works.

File: src/reducer.ts [hunk 8]
```diff
@@ -108,50 +139,120 @@ type ReducerCreator = {
...
  *
- * This operator does not change whether the stream is hot or cold.
+ * The behaviour is undefined if multiple reducers are registered for the same
+ * actions.
+ *
+ * This operator will only emit on relevant actions or if one of the source
+ * streams emits. Specifically, if the seed is a stream, this operator will not
+ * emit when the seed stream emits, but the first time an action of source
+ * stream emits after the seed stream has emitted. The seed stream will be
+ * unsubscribed after its first emit.
```
Comment (chriskr) [inline src/reducer.ts]: At the moment I don't se the use case of seed as stream, do we already use this in our app?

This 

> Specifically, if the seed is a stream, this operator will not
   emit when the seed stream emits, but the first time an action of source
   stream emits after the seed stream has emitted.

Seems not entirely trivial, what is the motivation?

File: src/reducer.ts [hunk 9]
```diff
@@ -108,50 +139,126 @@ type ReducerCreator = {
...
-        // This shouldn't be possible
-        return state;
-      }
+    ofType(...actionReducers.flatMap(reducerFn => reducerFn.trigger.actions)),
+    map((action): Packet => ({ source: ACTION_ORIGIN, value: action })),
+    merge(seed$, ...source$s),
+    scan(
+      (accumulator, packet) => {
+        if (packet.source === SEED_ORIGIN) {
+          return { initialized: true, reduced: false, state: packet.value };
```
Comment (chriskr) [inline src/reducer.ts]: `reduced` is no longer needed?

Comment (totokaka) [reply src/reducer.ts]: Good spot

**Comment by chriskr** [comment]
> 1. Agreed. I've removed this check, so now the seed will be emitted once it's ready. This is a breaking change when seed is not a stream, as previously that would not be emitted

I took a closer look at the table view. Currently nextState is always called twice when opening that view, one from the kickstart and one from the combined sources stream. I think that is not ideal, we saw that nextState can be more expensive than expected. For most of the views we took care that valid state gets calculated at the moment of connecting the view, it seems now we will do this twice?

**Comment by totokaka** [comment]
> > 1. Agreed. I've removed this check, so now the seed will be emitted once it's ready. This is a breaking change when seed is not a stream, as previously that would not be emitted
> 
> I took a closer look at the table view. Currently nextState is always called twice when opening that view, one from the kickstart and one from the combined sources stream. I think that is not ideal, we saw that nextState can be more expensive than expected. For most of the views we took care that valid state gets calculated at the moment of connecting the view, it seems now we will do this twice?

Whether we will emit multiple states right away depends on the reducers. In the table view I think we have the issue that our data reducers will emit on subscription, because the streams they depend on have `shareReplay`. This means that once the reducer is "attached" the streams will emit a value and the reducer will be invoked. 

We can solve this in three ways, I think:
  1. Not emit the seed from `combineReducers`
  2. Skip the replayed elements from the source streams of the reducers
  3. Debounce after `combineReducers`

Now, from the review, and in my opinion, `combineReducers` should emit the seed, as that seems most intuitive. Debouncing the stream means the state will still be calculated, but not  emitted, which is a workaround. So I guess the most apropriate solution is to skip the replayed elements on the reducers.

**Comment by anton164** [comment]
Removing myself as reviewer for now. Feel free to re-request it's ready with the changes we discussed

File: src/reducer.spec.ts [hunk 10]
```diff
@@ -39,35 +46,94 @@ test('reducer should store reducer function', t => {
...
+      expected$
+    );
+  })
+);
+
+test(
+  'combineReducers should reduce in predictable order',
+  marbles(m => {
+    // The "predictable order" seems to be the order in which the streams emit.
+    // Eg. The ordering of these seed definitions matter, but not the order of
```
Comment (anton164) [inline src/reducer.spec.ts]: I'm quite surprised to see that the order of the stream definitions matter 🤔

**Pull request merged**